baselines/deep_learning/util.py

```diff
@@ -0,0 +1,201 @@
+
+# coding=utf-8
+import torch
+from torch.nn import functional as F
+from torch.nn.modules import Module
+from torch.utils.data import DataLoader
+from Model import Protonet
+from Datagenerator import Datagen_test
+import numpy as np
+from batch_sampler1 import EpisodicBatchSampler
+from tqdm import tqdm
+from scipy.ndimage.filters import maximum_filter1d, minimum_filter
+from torch.autograd import Variable
+
+
+
+
+def euclidean_dist(x, y):
+    '''
+    Compute euclidean distance between two tensors
+    '''
+    # x: N x D
+    # y: M x D
+    n = x.size(0)
+    m = y.size(0)
+    d = x.size(1)
+    if d != y.size(1):
+        raise Exception
+
+    x = x.unsqueeze(1).expand(n, m, d)
+
+    y = y.unsqueeze(0).expand(n, m, d)
+
+    return torch.pow(x - y, 2).sum(2)
+
+
+
+def prototypical_loss(input, target,n_support):
+    '''
+    Inspired by https://github.com/jakesnell/prototypical-networks/blob/master/protonets/models/few_shot.py
+    Compute the barycentres by averaging the features of n_support
+    samples for each class in target, computes then the distances from each
+    samples' features to each one of the barycentres, computes the
+    log_probability for each n_query samples for each one of the current
+    classes, of appartaining to a class c, loss and accuracy are then computed
+    and returned
+    Args:
+    - input: the model output for a batch of samples
+    - target: ground truth for the above batch of samples
+    - n_support: number of samples to keep in account when computing
+      barycentres, for each one of the current classes
+    '''
+    target_cpu = target.to('cpu')
+    input_cpu = input.to('cpu')
+    n_query=n_support
+
+    classes = torch.unique(target_cpu)
+    n_classes = len(classes)
+    p = n_classes * n_support
+
+    prototypes = input_cpu[:p].view(n_support, n_classes, -1).mean(0)
+    query_set = input_cpu[p:]
+
+    dists = euclidean_dist(query_set, prototypes)
+
+
+    log_p_y = F.log_softmax(-dists, dim=1).view(n_classes, n_query, -1)
+
+    target_inds = torch.arange(0, n_classes)
+    target_inds = target_inds.view(n_classes, 1, 1)
+    target_inds = target_inds.expand(n_classes, n_query, 1).long()
+    loss_val = -log_p_y.gather(2, target_inds).squeeze().view(-1).mean()
+    _, y_hat = log_p_y.max(2)
+
+    acc_val = y_hat.eq(target_inds.squeeze()).float().mean()
+    return loss_val, acc_val
+
+
+def get_probability(x_pos,neg_proto,query_set_out):
+
+
+    """Calculates the  probability of each query point belonging to either the positive or negative class
+     Args:
+     - x_pos : Model output for the positive class
+     - neg_proto : Negative class prototype calculated from randomly chosed 100 segments across the audio file
```

Comment:
Is this still from 100 segments? It seems like it might be too many?

Reply:
Would it be too much to have multiple negative prototypes. that each is compute from 5 samples. and then compute the distance from each of them and the distance from the positive and if get the probabilities etc like normal?